design-proposals/configurable-features.md

```diff
@@ -0,0 +1,229 @@
+[...]
+status:
+  conditions:
+  - lastProbeTime: "2024-08-28T10:16:57Z"
+    lastTransitionTime: "2024-08-28T10:16:57Z"
+    message: virtual machine is requesting the disabled feature: feature-B 
+    reason: FeatureNotEnabled
+    status: "False"
+    type: Synchronized
+```
+
+and a warning event is triggered:
+
+```event
+LAST SEEN   TYPE      REASON                    OBJECT                                 MESSAGE
+[...]
+2s          Warning   FeatureNotEnabled         virtualmachineinstance/vmi-feature-b   feature-B feature not enabled
+```
+
+- Feature configuration checks that could prevent a VMI from starting should only be performed during the VMI
+  reconciliation process, and not at runtime if the changes cannot be applied without restarting the VMI. While this
+  approach ensures that the system does not actively block, stop, or kill running VMIs due to configuration changes in
+  the KubeVirt CR, it is important to note that VMIs may still experience issues or termination if critical features
+  become unavailable or incompatible. 
+- The system should not block live migration unless the requested feature
+  is not supported in the destination host. However, as stated before, if the changes can be applied without 
+  restarting VMI, it can be done at runtime.
+- Updates to KubeVirt CR to update a feature configuration should not be rejected.
+
+## Scalability
+
+The feature configurables should not affect in a meaningful way the cluster resource usage.
+
+## Update/Rollback Compatibility
+
+The feature configurables should not affect forward or backward compatibility once the feature GA. A given feature,
+after 3 releases in Beta, all feature gates must be dropped. Those features that need a configurable should define it ahead
+of time.
+
+## Functional Testing Approach
+
+The unit and functional testing frameworks should cover the relevant scenarios for each feature.
+
+# Implementation Phases
+
+The feature configuration checks should be placed in the VMI reconciliation loop. In this way, the feature configuration
+evaluation is close to the VMI scheduling process, as well as allowing KubeVirt to reconcile itself if it is out of sync
+temporally.
+
+Regarding already existing features transitioning from feature gates as a way to enable/disable a feature to configurable
+fields, this change is acceptable, but it should be marked as a breaking change and documented. Moreover, all feature
+gates should be evaluated to determine if they need to be dropped and transitioned to configurables.
+
+## About implementing the checking logic in the VM controller
+
+The checking in the VM controller could be added to let the user know if a VM has requested a feature configuration which 
+is different from what it is specified in the KubeVirt CR. This will provide an early feedback to the user before starting
+the VM. However, it should not prevent the user to start the VM, the VMI controller should take care of checks preventing
```

Comment:
I don't understand the English here, e.g what is the "early feedback" you refer to.

Can you rephrase the paragraph to ensure that
* The controller must not start a VM that requires a feature that is not available in the cluster.
* The fact that a VM cannot start despite its owner asking to do so must be reported by the controller in the `status` field of the VM.

Reply:
Done.
Let me clarify what I wanted to express with "early feedback". Let's suppose that the `feature-b` is disabled cluster-wide and the user creates a VM object requesting this feature. The VM is not requested to start, just the VM object is created. The idea is that the VM controller updates the `status` field reflecting that this VM won't start because it is requesting a feature no available.
I'm not sure if this would be helpful for users or just will create more noise.

Reply:
Ah. You would like to inform a stopped VM that it won't be able to generate a VMI. This can be quite expensive (as it requires you to track all stopped VMs for no immediate reason). I would not make this a requirement.

Reply:
AFAIK, it is not that expensive. It's a check in the VM controller reconciliation loop, what it is true is that if you have a stopped VM and disable the feature, the reconciliation loop is not triggered immediately. Takes some time. I will make this optional.

Reply:
Done. Could you please add the lgtm again if you are fine with the change? Thanks.